cpp/src/stream_compaction/distinct_helpers.cu

```diff
@@ -22,118 +22,105 @@
-                       ref.fetch_max(idx, cuda::memory_order_relaxed);
-                     } else {
-                       // Count the number of rows in each group of rows that are compared equal.
-                       ref.fetch_add(size_type{1}, cuda::memory_order_relaxed);
-                     }
-                   });
-
-  auto const map_end = [&] {
-    if (keep == duplicate_keep_option::KEEP_NONE) {
-      // Reduction results with `KEEP_NONE` are either group sizes of equal rows, or `0`.
-      // Thus, we only output index of the rows in the groups having group size of `1`.
-      return thrust::copy_if(
-        rmm::exec_policy(stream),
-        thrust::make_counting_iterator(0),
-        thrust::make_counting_iterator(num_rows),
-        output_indices.begin(),
-        cuda::proclaim_return_type<bool>(
-          [reduction_results = reduction_results.begin()] __device__(auto const idx) {
-            return reduction_results[idx] == size_type{1};
-          }));
-    }
+  auto keys   = rmm::device_uvector<size_type>(num_rows, stream);
+  auto values = rmm::device_uvector<size_type>(num_rows, stream);
+
+  auto pairs = thrust::make_transform_iterator(
+    thrust::counting_iterator<size_type>(0),
+    cuda::proclaim_return_type<cuco::pair<size_type, size_type>>([] __device__(size_type const i) {
+      return cuco::pair<size_type, size_type>{i, 1};
+    }));
+
+  map.insert_or_apply_async(pairs, pairs + num_rows, plus_op{}, stream.value());
+  auto const [keys_end, _] = map.retrieve_all(keys.begin(), values.begin(), stream.value());
+
+  auto num_distinct_keys = thrust::distance(keys.begin(), keys_end);
+  keys.resize(num_distinct_keys, stream);
+  values.resize(num_distinct_keys, stream);
+
+  auto output_indices = rmm::device_uvector<size_type>(num_distinct_keys, stream, mr);
+
+  auto const output_iter = cudf::detail::make_counting_transform_iterator(
+    size_type(0),
+    cuda::proclaim_return_type<size_type>(
+      [keys = keys.begin(), values = values.begin()] __device__(auto const idx) {
+        return values[idx] == size_type{1} ? keys[idx] : -1;
+      }));
 
-    // Reduction results with `KEEP_FIRST` and `KEEP_LAST` are row indices of the first/last row in
-    // each group of equal rows (which are the desired output indices), or the value given by
-    // `reduction_init_value()`.
-    return thrust::copy_if(
-      rmm::exec_policy(stream),
-      reduction_results.begin(),
-      reduction_results.end(),
-      output_indices.begin(),
-      cuda::proclaim_return_type<bool>([init_value = reduction_init_value(keep)] __device__(
-                                         auto const idx) { return idx != init_value; }));
-  }();
+  auto const map_end = thrust::copy_if(
+    rmm::exec_policy_nosync(stream),
+    output_iter,
+    output_iter + num_distinct_keys,
+    output_indices.begin(),
+    cuda::proclaim_return_type<bool>([] __device__(auto const idx) { return idx != -1; }));
```

Comment:
Please add back the comment.  Although I wrote the original code, now I hardly understand what it is doing without the comment.